src/Npgsql/NpgsqlDataReader.cs

```diff
@@ -1391,6 +1697,218 @@ public override DataTable GetSchemaTable()
 
         #endregion Schema metadata table
 
+        #region Seeking
+
+        Task SeekToColumn(int column, bool async)
+        {
+            if (_isSequential)
+                return SeekToColumnSequential(column, async);
+            else
+            {
+                SeekToColumnNonSequential(column);
+                return PGUtil.CompletedTask;
+            }
+        }
+
+        void SeekToColumnNonSequential(int column)
+        {
+            if (_column == -1)
+            {
+                // Starting a new row.
+
+                // We currently assume that the row's number of columns is identical to the description's
+#if DEBUG
+                _numColumns = Buffer.ReadInt16();
+                Debug.Assert(RowDescription.NumFields == _numColumns);
+#else
+                Buffer.ReadPosition += 2;
+#endif
+
+                // Initialize our columns array with the offset and length of the first column
+                _columns.Clear();
+                var len = Buffer.ReadInt32();
+                _columns.Add((Buffer.ReadPosition, len));
+            }
+
+            // Shut down any streaming going on on the column
+            if (ColumnStream != null)
+            {
+                ColumnStream.Dispose();
+                ColumnStream = null;
+                // Disposing the stream leaves us at the end of the column
+                PosInColumn = ColumnLen;
+            }
+
+            for (var lastColumnRead = _columns.Count; column >= lastColumnRead; lastColumnRead++)
+            {
+                int lastColumnLen;
+                (Buffer.ReadPosition, lastColumnLen) = _columns[lastColumnRead-1];
+                if (lastColumnLen != -1)
+                    Buffer.ReadPosition += lastColumnLen;
+                var len = Buffer.ReadInt32();
+                _columns.Add((Buffer.ReadPosition, len));
+            }
+
+            (Buffer.ReadPosition, ColumnLen) = _columns[column];
+            _column = column;
+            PosInColumn = 0;
+        }
+
+        /// <summary>
+        /// Seeks to the given column. The 4-byte length is read and stored in <see cref="ColumnLen"/>.
+        /// </summary>
+        async Task SeekToColumnSequential(int column, bool async)
+        {
+            if (_column == -1)
+            {
+                // Starting a new row.
+
+                // We currently assume that the row's number of columns is identical to the description's
+                await Buffer.Ensure(2, async);
+#if DEBUG
+                _numColumns = Buffer.ReadInt16();
+                Debug.Assert(RowDescription.NumFields == _numColumns);
+#else
+                Buffer.ReadPosition += 2;
+#endif
+            }
+
+            if (column < 0 || column >= _numColumns)
+                throw new IndexOutOfRangeException("Column index out of range");
```

Comment:
`ArgumentOutOfRangeException` fits better here.

Reply:
This is what [the docs specify we should throw](https://docs.microsoft.com/en-us/dotnet/api/system.data.common.dbdatareader.item?view=netframework-4.7.2#System_Data_Common_DbDataReader_Item_System_Int32_), I also confirmed that this is the SqlClient behavior.

Reply:
It`s surprisingly unexpected to me to throw this kind of exception for methods like `GetFieldValue<T>`. Maybe it's because these methods were introduced after `IDataReader` and its indexer properties. Thanks for pointing out!

Reply:
I understand... I think the idea is that the reader exposes an array-like interface - so going out of bounds should throw the same thing as an array. But in any case we have the docs...